### ISSUE TITLE
Recommended API to set Generator for module?

### ISSUE THREAD

**Issue by allefeld** [numpy/numpy#15571]
If I develop a module that uses random numbers from `numpy.random`, and follow the recommendation to use `numpy.random.default_rng()` to create a generator, e.g. in the module's initialization code, then every time the module is freshly imported I get a new sequence of random numbers.

Now it is often useful, for example for running exactly the same simulation several times, to initialize the generator to a specific state, and this creation function provides that possibility. I can do that for testing purposes during development, but in the final version I would of course use the form in which "fresh entropy" is used.

Now, if someone else were to use my module, and they want exactly the same behavior on multiple runs, e.g. for testing their code which builds on mine, they would not have the possibility to enforce that for my module short of editing my code. I think it would be better to offer such a user an API, such that in normal use a "fresh entropy" generator is created, but it is also possible to set a specific one. I believe that is what "passing around a generator" refers to in [NEP 19](https://numpy.org/neps/nep-0019-rng-policy.html).

Finally my question: Is there a recommended way to provide such an API? One possibility would have to have a consistently named module-global variable (e.g. `__rng__`), which could be read to use the same generator elsewhere, or set to force the module using another one.

I understand the concerns that lead to avoid having a "default global instance". But the drawback is what I described above: for a program consisting of different pieces all using `numpy.random`, but from different authors, it is not easily possible to enforce a specific RNG generator seed.

**Comment by rkern** [comment]
When I refer to "passing around a generator", I mean that it should be an explicit argument in all of the functions that use the generator. If your function calls another function that uses the generator, it should pass that along in the arguments. There should not be any globals, module-level or otherwise. This is the way that code from multiple authors work together.

`scikit-learn` and `scipy` do this very effectively using a `check_random_state()` utility function (for the old `RandomState` system). You can use `default_rng()` more or less the same way.

```python
def f(x, y, rng=None):
    rng = np.random.default_rng(rng)
    z = rng.normal(x, y)
    return g(z, rng)

def g(z, rng=None):
    rng = np.random.default_rng(rng)
    return z + rng.uniform(0, 10)
```

If I call `f(x, y)`, it will use a `Generator` with fresh entropy. If I call `f(x, y, rng=1234567890)`, it will use a `Generator` seeded with `1234567890`. If I call `f(x, y, rng=some_generator_i_have_from_somehwere)`, it will use that `Generator` instance.

**Comment by allefeld** [comment]
Thanks!

**Comment by mattip** [comment]
@rkern should we reopen this as documentation issue or does that best practices recommendation already explicitly exist soemwhere and I just don't see it?

**Comment by allefeld** [comment]
I have to say that the best documentation is often to be found in issues. ;-) But I agree, having that stated explicitly would be good.

I suggest to expand on "This function does not manage a default global instance." under <https://docs.scipy.org/doc/numpy/reference/random/generator.html>. At least that's the place where I got stuck, then found NEP 19 when googling "numpy.random reproducible". That text is quite useful, so I suggest linking to it in this place, too.

**Comment by mattip** [comment]
Reopening as a documentation issue.

**Comment by josesho** [comment]
Second this! 

It would also be helpful to understand what exactly the differences are between `RandomState`  and `Generator`, because both methods do not give the same results even if they look the same.

```python
import numpy as np
from numpy.random import Generator, RandomState, SeedSequence, PCG64

ss = SeedSequence(12345)

pcg_rs = RandomState(PCG64(ss))
pcg_rs.standard_normal(2)
```
gives

```shell
array([-0.72298999, -1.07581063])
```

but 

```python
pcg_gen = Generator(PCG64(ss))
pcg_gen.standard_normal(2)
```

gives
```shell
array([-1.42382504,  1.26372846])
```

🤷🏻‍♂️

**Comment by rossbar** [comment]
@josesho In my opinion, the [random docs](https://numpy.org/doc/1.18/reference/random/index.html#random-sampling-numpy-random) are already pretty explicit on this topic. From the overview:

> The legacy RandomState random number routines are still available, but limited to a single BitGenerator.

It is not possible to create a `RandomState` instance with the `PCG64` BitGenerator as `RandomState` is the *legacy* interface and has the Mersenne twister baked into it. In fact, this is one of the primary motivations for updating `random`.

In my opinion, the docs make this distinction quite clear, but if there are specific places in the docs that you think can be improved, please don't hesitate to open a PR with proposed changes!

**Comment by seberg** [comment]
Except that it is possible to create a `RandomState` with a PCG64, which is not really something we want users to do. Should we disallow specifying a bitgenerator as initializer argument to randomstate @bashtage? It should be discouraged, but I do not know if this is just possible due to them sharing the implementation, or if it was intentional?

**Comment by bashtage** [comment]
I think it has to be the case a bit generator can be given the RandomState since this allows the same source of pseudo-random bits to be shared across a RandomState and a Generator.  For example, suppose you need legacy normals but want to use one of @WarrenWeckesser recent additions that only exists in `Generator`
```
entropy = 0xFEEDC0DE
from numpy.random import MT19937, RandomState, Generator

mt = MT19937(entropy)
rs = RandomState(mt)
rg = Generator(mt)
```

It could be the case that `RandomState` is restricted to `MT19937`, although I don't see an overwhelming gain from imposing this.

**Comment by rossbar** [comment]
Whoops, then the docs do seem to be a bit out of sync with the behavior - the very top line in the `np.random.RandomState` doc string:

> Init signature: np.random.RandomState(self, /, *args, **kwargs)
> Docstring:     
> RandomState(seed=None)
> 
> Container for the slow Mersenne Twister pseudo-random number generator.
> Consider using a different BitGenerator with the Generator container
> instead.

**Comment by charris** [comment]
IIRC, Random state has the Mersenne Twister built into it's methods and cannot be used with the other bit generators.

**Comment by rossbar** [comment]
@bashtage If I understand your comment correctly - I believe the following can be done to get legacy normals with the Generator *without* having to pass the BitGenerator to RandomState:

```python
rs = RandomState(entropy)
mt = MT19937()
mt.state = rs.get_state()
rg = Generator(mt)
rvals = rg.normal(1000)
```
It's clunkier than your example (and maybe not recommended?), but it does seem to be possible without passing the BitGenerator to RandomState

**Comment by bashtage** [comment]
That won't keep everything in sync with a single steam since you have two
distinct bit generators.  In fact, you would use the same bit sequences
across the two which would generate high depended between variates.

**Comment by rkern** [comment]
> IIRC, Random state has the Mersenne Twister built into it's methods and cannot be used with the other bit generators.

No, `RandomState` does accept any `BitGenerator`, as demonstrated in [this comment](https://github.com/numpy/numpy/issues/15571#issuecomment-589545055). This was one of the explicit design requirements.

https://numpy.org/neps/nep-0019-rng-policy.html#supporting-unit-tests

The difference noted in [the comment](https://github.com/numpy/numpy/issues/15571#issuecomment-589545055) has to do with the fact that we updated the `standard_normal()` algorithm in `Generator` to use the Ziggurat method.

**Comment by rkern** [comment]
> I believe the following can be done to get legacy normals with the Generator without having to pass the BitGenerator to RandomState

No, because we use a different algorithm for normals in `Generator` than `RandomState`, regardless of the underlying uniform bit-stream.

**Comment by rkern** [comment]
These differences are pretty prominently documented: https://docs.scipy.org/doc/numpy/reference/random/index.html#what-s-new-or-different

**Comment by rossbar** [comment]
> No, because we use a different algorithm for normals in Generator than RandomState, regardless of the underlying uniform bit-stream.

Good point, I had forgotten about that, so `normal` was a bad example. For uniform random numbers, the example I highlighted above does seem to work, though perhaps it's fragile - I'm not familiar enough with `random` fully understand [bashtage's comment](https://github.com/numpy/numpy/issues/15571#issuecomment-589794328).
```python
import numpy as np
from numpy.random import MT19937, Generator, RandomState
entropy = 0xFEEDC0DE
rs = RandomState(entropy)
mt = MT19937()
mt.state = rs.get_state()
rg = Generator(mt)
legacy_vals = rs.random(1000000)
generator_vals = rg.random(1000000)
np.allclose(legacy_vals, generator_vals) #True
```

From my perspective (as someone who is trying to learn about the `random` updates from the docs) there is a lot of great info but it is difficult to assimilate it all. I'm not sure what actionable items can be derived from the discussion, though I very much appreciate all the clarification from @rkern and @bashtage

**Comment by rkern** [comment]
@bashtage was demonstrating the use case that motivated why we modified the legacy `RandomState` to accept the new `BitGenerator`s instead of just leaving `RandomState` completely alone and fixed to MT19937. This motivation is alluded to in the section of NEP 19 that I linked above. He's not saying that there aren't other ways to instantiate `RandomState` and `Generator` that would create identical uniform streams.

**Comment by WeirdKeksButtonK** [comment]
I have a related question, not sure if it is worth opening a new issue, so asking here.

I understand that the best way to work with randomness in the new style is adding the `rng` argument everywhere. Now, let's say I am developing a micro-library called `rndm` with the only goal to simplify usage of the new style `np.random` by having a global `rng` instance and adding a possibility to set its state. In other words, the library intentionally doesn't follow the best practices and tries to take the best of the two worlds: new Generator and old ergonomics.

```python
# rndm/__init__.py
from np.random import default_rng

rng = default_rng()

def set_rng(value):
    global rng
    rng = default_rng(value)
```
Usage (see [this comment](https://github.com/numpy/numpy/issues/15571#issuecomment-586609599)):
```python
from rndm import rng

def f(x, y):
    z = rng.normal(x, y)
    return g(z)

def g(z):
    return z + rng.uniform(0, 10)
```

Now, I am trying to understand if ergonomics outweigh potential problems caused by `rndm` in my scenarios, so the questions are:
1. What problems the library `rndm` may cause for its users?
2. Are those problems different from the problems that the old style `np.random` has?

**Comment by mattip** [comment]
The subtle problems the new API solves are:
- Repeatability: when you create a new seeded `Generator(seed)` locally, you can be certain the random sequence you generate will be repeatable in the local context, independent of global sate. This is not guaranteed by your scheme since changing the order or number of times you call `f` and `g` or addng a function `h` which also uses `rndm` would change the results.
- Getting independent streams when called in [different threads or processes](https://numpy.org/doc/stable/reference/random/parallel.html). The new API has a mechanism to spawn new `SeedSequences` which will ensure that each one is, for all practical purposes, independent. Using a global `Generator` could cause your streams to be correlated, creating biases in your random sequences.

The code in the comment you link to at least provides a way to mitigate these by allowing you to pass in a seed. For tests, you may choose to use a fixed seed (1234567890 in the example) to ensure your results are constant, and for code that operates in parallel you could pass in one of the spawned `SeedSequences`.

**Comment by bashtage** [comment]
> I understand that the best way to work with randomness in the new style is adding the `rng` argument everywhere. Now, let's say I am developing a micro-library called `rndm` with the only goal to simplify usage of the new style `np.random` by having a global `rng` instance and adding a possibility to set its state. In other words, the library intentionally doesn't follow the best practices and tries to take the best of the two worlds: new Generator and old ergonomics.

I wouldn't do it this way. Probably something like

```
"""
rndm.py
"""

from numpy.random import default_rng

_gen = default_rng()
_bit_gen = _gen.bit_generator
_state = _bit_gen.state

def set_state(state):
    _bit_gen.state = state

def get_state(state):
    return _bit_gen.state

def reset_state():
    _gen.state = _state

standard_normal = _gen.standard_normal
<lots of other functions>
...
```

**Comment by WeirdKeksButtonK** [comment]
> I would probably use something like:

I'm not sure that this approach actually works. The function `standard_normal` is a bound method of the first `_gen` instance and will not be affected by the `seed(...)` call. So you have to reset all the functions as well.

Actually, there is a similar problem with the snippet I suggested above: it is _very_ easy to misuse the global instance by importing it directly like `from rndm import rng`, because the `rng` variable will not be affected by the `set_rng(...)` call. There are possible workarounds for this issue as well, however, I decided to try to live with the new "pass rng as an argument" paradigm :)

**Comment by bashtage** [comment]
You are right.  This fixes it

```
# or reseed
def seed(seed_val: Union[int, Sequence[int], SeedSequence]):
    _gen.state = default_rng(seed_val).bit_generator.state
```

**Comment by endolith** [comment]
@mattip 

> For tests, you may choose to use a fixed seed (1234567890 in the example) to ensure your results are constant

I can't figure out how to do this.  Previously a unit test would start with `np.random.seed(1234)` to restart the generator at the same place each time, but how do we do this with the new `default_rng` if there's no way to pass the generator object into the functions being tested?

**Comment by bashtage** [comment]
You can use

```python
def test_something():
    rg = np.random.default_rng(FIXED_SEED)
    ...
```

or if you are using pytest

```python
@pytest.fixture(scope="function")
def rg():
    return np.random.default_rng(FIXED_SEED)

def test_something(rg):
    rg.standard_normal(...)
```

**Comment by mattip** [comment]
> This approach may be a bit brittle if you try and run tests in parallel ...

... or if you use any technique that changes the sequence of calls to the global generator instance:
- any optional test skips or xfails for certain scenarios/platforms/hardware availability
- rerunning tests but choosing only certain ones with command line `pytest ... -k <test-to-run>
- developing and commenting out a test temporarily
...

**Comment by bashtage** [comment]
> ... or if you use any technique that changes the sequence of calls to the global generator instance:
> 
> * any optional test skips or xfails for certain scenarios/platforms/hardware availability
> * rerunning tests but choosing only certain ones with command line `pytest ... -k
> * developing and commenting out a test temporarily

I don't think it is that brittle. The only scenario I can imagine where you would get the wrong sequence would be if two threads were using the same Generator instance at the same time.  Most test runners using multiprocessing and so would not break.  I still think it is unnecessarily complex when compared to the first or second method, and there is literally no gain to it. I only included it because it is the closest thing to seeding the singleton generator (which has the same potential brittleness).

**Comment by mattip** [comment]
Just to be clear, since my comment was too terse: I was referring to the usage pattern of instantiating a global singleton seed sequence (or singleton generator with a seed sequence) and using it across multiple tests. This is a bad usage pattern for reproducibility, and one of the major advantages of the new API over the old one. Anything that changes the order of tests will mess with the stream that any particular test sees.

**Comment by endolith** [comment]
@bashtage 


> You can use
> 
> ```python
> def test_something():
>     rg = np.random.default_rng(FIXED_SEED)
>     ...
> ```


That would be fine if the test were something like:

```py
def test_processor1():
    np.random.seed(1234)
    data = np.random.rand(100)
    assert_allclose(processor1(data), known_good)

def test_processor2():
    np.random.seed(1234)
    data = np.random.rand(100)
    assert_allclose(processor2(data), known_good)
```

as is currently recommended by https://github.com/numpy/numpy/blob/master/doc/TESTS.rst.txt#tests-on-random-data

This would become:

```py
def test_processor1():
    rg = np.random.default_rng(1234)
    data = rg.random(100)
    assert_allclose(processor1(data), known_good)

def test_processor2():
    rg = np.random.default_rng(1234)
    data = rg.random(100)
    assert_allclose(processor2(data), known_good)
```

but I'm asking about a case like this:

```py
def test_generator():
    np.random.seed(1234)
    data = generating_func(100)
    assert_allclose(data.mean(), known_good)
```

 `rg` has no effect on `generating_func`.  How to reset the RNG inside `generating_func` for each test function, to ensure that failures are only coming from bugs in `generating_func` or changes to the test itself, and not from random variation exceeding the tolerances of `assert_allclose`?  (Which creates false negatives all the time in scipy, causing obscure tests to fail when the PR has not actually broken anything.)

**Comment by bashtage** [comment]
If `generating_func` relies on the singleton generator (that should be considered a bug, but I digress), then yes, you need to set the singleton seed as you are going above.  

```python
def test_generator():
    np.random.seed(1234)
    data = generating_func(100)
    assert_allclose(data.mean(), known_good)
```

If `generating_func` is fixed to accept a `Generator` instance, then 

```python
def test_generator():
    rg = np.random.default_rng(1234)
    data = generating_func(100, random_state=rg) # like scikit-learn and scipy IIRC
    assert_allclose(data.mean(), known_good)
```

**Comment by endolith** [comment]
> If `generating_func` relies on the singleton generator (that should be considered a bug, but I digress), 

It previously did, but I changed it to use `default_rng` internally, as in the directions on https://numpy.org/doc/stable/reference/random/index.html#quick-start

> If `generating_func` is fixed to accept a `Generator` instance, then

Ah, ok, so all functions that generate random numbers internally must be changed to accept a `Generator` instance.  That would be good to explain in  https://numpy.org/doc/stable/reference/random/index.html#quick-start

So "Use SeedSequence.spawn" in the table isn't relevant?

**Comment by rkern** [comment]
> Ah, ok, so all functions that generate random numbers internally must be changed to accept a `Generator` instance. That would be good to explain in https://numpy.org/doc/stable/reference/random/index.html#quick-start

That's what the first sentence says: "Call `default_rng()` to get a new instance of a `Generator`, then call its methods to obtain samples from different distributions."

**Comment by afonari** [comment]
Hi to all!

Can someone clarify if it is possible to save Generator's state, similar to the RandomState.get_state function?

It is important for checkpoint systems for example, to be able to dump the generator state in json (or whatever) and pick it up from there.

Thanks!

**Comment by bashtage** [comment]
BitGenerators have state, generators do not.  You get it from the bit generator.

```
gen = np.random.default_rng(20210318)
state = gen.bit_generator.state
print(gen.integers(10,size=10))

gen.bit_generator.state = state
print(gen.integers(10,size=10))
```

which prints

```
[7 9 1 3 1 6 7 3 4 5]
[7 9 1 3 1 6 7 3 4 5]
```

**Comment by afonari** [comment]
Ah, so there are couple of bit generators: MT19937, PCG64, Philox, SFC64 (https://numpy.org/doc/stable/reference/random/bit_generators/index.html). And each of them has state getter and setter which is a dict. Gotcha, thanks.

**Comment by bashtage** [comment]
There are a wider range of bit generators available in the randomgen package:

https://bashtage.github.io/randomgen/bit_generators/index.html

**Comment by hugobuddel** [comment]
Perhaps the documentation of `np.random.seed()` should be updated as well. I figured that my use of `seed()` is sub-optimal, and help() confirmed my suspicion. But the best practice example did not tell me enough to figure out how to do better:
```
>>> help(np.random.seed)
Help on built-in function seed:

seed(...) method of numpy.random.mtrand.RandomState instance
    seed(self, seed=None)
    
    Reseed a legacy MT19937 BitGenerator
    
    Notes
    -----
    This is a convenience, legacy function.
    
    The best practice is to **not** reseed a BitGenerator, rather to
    recreate a new one. This method is here for legacy reasons.
    This example demonstrates best practice.
    
    >>> from numpy.random import MT19937
    >>> from numpy.random import RandomState, SeedSequence
    >>> rs = RandomState(MT19937(SeedSequence(123456789)))
    # Later, you want to restart the stream
    >>> rs = RandomState(MT19937(SeedSequence(987654321)))
```

From just that example it is not clear how one would proceed to get non-random random numbers, because the example is not a drop-in replacement: `np.random.random()` still gives different numbers, unlike when using `np.random.seed()`.

Maybe something like the following should be added?:
```
    rs can subsequently be used to draw pseudo-random numbers from.

    >>> rs.random()
    0.16693771389729373
```

(This is not a complaint: I'm very happy that it is easier to do this well now!)

**Comment by mattip** [comment]
We should also note that `RandomState` is the older API, and new code should be using `rng = Generator(PCG64(123456789))`.
